### DISCUSSION TITLE
How to replace ipxe inside netboot.xyz

### DISCUSSION THREAD

**Discussion by lars18th** [antonym/netboot.xyz#599] [General]
Hi,

I hope this will not be ofensive. I'm new to netboot.xyz, and I'm still reading documentation about this project.

The question is simple: I need to "replace" the ipxe inside the netboot.xyz image (ISO, IMG or PXE). The reason is becuase some drivers that I need aren't compiled by default (rndis, acm, ecm, ncm, etc.). So my idea is to compile (make) a custom version of `ipxe.krn` and replace it inside the ISO image (IMG or PXE). However, I don't know how to compile correctly the ipxe version used inside this project.

Please, can you point me to the correct documentation/howto/buid_script to complete this task?
Thank you!

**Comment by thelamer** [comment]
You can use ansible to compile new bootloader files , here is the actual make command https://github.com/netbootxyz/netboot.xyz/blob/development/roles/netbootxyz/tasks/generate_disks_legacy.yml#L26
 Basic build instructions are here https://netboot.xyz/selfhosting/

**Reply by lukemilby** [reply]
for the next person trying to compile a new bootloader https://netboot.xyz/docs/selfhosting

**Reply by lars18th** [reply]
Hi @lukemilby ,

But this guide doens't explain with sufficient info how to compile custom drivers and include them in the custom package.

**Comment by lars18th** [comment]
Hi @thelamer ,

Thank you for the info. I do the "ansible" compilation after editing the file pointed `generate_disks_legacy.yml` to include the compilation of the drivers that I need.

However, I found some troubles:

1. When changing the name of the "target", for instance from `bin/ipxe.dsk` to `bin/ipxe--rndis--acm--ecm--ncm--axge.dsk` (and so on), the generation of the package stops after the compilation of the ipxe source because the expected files are `ipxe.*`. Then I suggest to change the scripts to include _additional drivers_ and complete the generation of the package renaming the target to the short names.
2. Futhermore, I feel that at time it's quite complex to add more drivers to the booting medium. As more and more devices will appear _without_ internal Ethernet nics, I suggest to add some stock support for USB devices. I suggest to prepare a new target like "ipxe-usb-drivers" with all USB drivers included. If the previous issue will be fixed then it will be more easy to add such drivers.
3. Finally, I see that some of the hosted live distributions don't have native support to boot using USB network devices. I suggest to check this and configure them to boot when using USB nics.

I hope you want to consider this.
Until them I post here the files generated with support for booting with USB Nics:
[ipxe--rndis--acm--ecm--ncm--axge.zip](https://github.com/netbootxyz/netboot.xyz/files/4509653/ipxe--rndis--acm--ecm--ncm--axge.zip)


Regards.
